adobe/react-spectrum
# fix(#2101): TooltipTrigger: Tooltip should remain hidden in the DOM to provide an IDREF for aria-describedby on trigger element

## 🐛 Bug Report

TooltipTrigger: Tooltip should remain hidden in the DOM to provide an IDREF for aria-describedby on trigger element.

### 🤔 Expected Behavior

In order for the Tooltip information to be available to assistive technology when navigating elements using JAWS's virtual cursor or NVDA's browse mode, the Tooltip that displays as an overlay in a portal when the trigger is either hovered or focused should remain hidden in the DOM to provide an IDREF for `aria-describedby` on trigger element. The `triggerProps` returned by `useTooltipTrigger` should always include `aria-describedby` referencing the tooltip by `id`, regardless of whether the tooltip is visible in the portal or hidden in the DOM.

### 😯 Current Behavior

Currently, a tooltip is only rendered to the DOM when the TooltipTrigger `state.isOpen === true`. The tooltip and the `aria-describedby` attribute on its trigger get removed when `state.isOpen === false`. This means that the text of the tooltip is only announced when the user focuses or hovers the trigger, and is not available to screen reader users navigating using virtual cursor/browse mode.

### 💁 Possible Solution

Keep the Tooltip in the DOM with a `hidden` attribute when the tooltip is not shown.

### 🔦 Context

This is important for improving screen reader accessibility.

### 💻 Code Sample

See https://codepen.io/Moiety/pen/LaPvWy for a positive example in development for WAI-ARIA 1.2 Authoring Practices: https://w3c.github.io/aria-practices/#tooltip

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  @react-spectrum/tooltip@3.1.3, @react-aria/tooltip@3.1.2
| Browser          | All
| Operating System | All

Closes #2101 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue #2101](https://github.com/adobe/react-spectrum/issues/2101).

## 📝 Test Instructions:
1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/674b18a8de35c388d5f289a7be8a1b4b77d3561f/storybook/index.html?path=/story/tooltiptrigger--default
2. Verify TooltipTrigger examples still work as expected on hover and focus.
3. Using a Windows screen reader, like JAWS or NVDA, navigate to the TooltipTrigger button in browse/virtual cursor mode. Neither JAWS nor NVDA will announce the Tooltip text by default, however:
    - with JAWS, INSERT+TAB, or CAPS LOCK+TAB on laptops, announces the current object in virtual cursor mode, including the Tooltip,
    - and similarly with NVDA, INSERT+SHIFT+O, or CAPS LOCK+SHIFT+O on laptops, announces the current object in virtual cursor mode, including the Tooltip.
4. Open https://reactspectrum.blob.core.windows.net/reactspectrum/674b18a8de35c388d5f289a7be8a1b4b77d3561f/storybook/index.html?path=/story/actiongroup--with-tooltips and verify that keyboard navigation of items in the ActionGroup still works.

## 🧢 Your Project:

Adobe/Accessibility
